pkg/kubelet/status/status_manager.go

```diff
@@ -335,19 +345,26 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
 		oldStatus = &cachedStatus.status
 	}
 	status := *oldStatus.DeepCopy()
-	for i := range status.ContainerStatuses {
-		if status.ContainerStatuses[i].State.Terminated != nil {
-			continue
-		}
-		status.ContainerStatuses[i].State = v1.ContainerState{
-			Terminated: &v1.ContainerStateTerminated{
-				Reason:   "ContainerStatusUnknown",
-				Message:  "The container could not be located when the pod was terminated",
-				ExitCode: 137,
-			},
+
+	// once a pod has initialized, any missing status is treated as a failure
+	if hasPodInitialized(pod) {
+		for i := range status.ContainerStatuses {
+			if status.ContainerStatuses[i].State.Terminated != nil {
+				continue
+			}
+			status.ContainerStatuses[i].State = v1.ContainerState{
+				Terminated: &v1.ContainerStateTerminated{
+					Reason:   "ContainerStatusUnknown",
+					Message:  "The container could not be located when the pod was terminated",
+					ExitCode: 137,
+				},
+			}
 		}
 	}
-	for i := range status.InitContainerStatuses {
+
+	// all but the final suffix of init containers which have no evidence of a container start are
+	// marked as failed containers
+	for i := range initializedContainers(status.InitContainerStatuses) {
```

Comment:
This is tricky.  The presence of an init container at `N+1` implies that any previous init container exited with code 0 at least one time.  Therefore by inductive reasoning, it would be wrong to transform `[<uninit>, exit=0] -> [exit=137, exit=0]`.  *Technically* `[<uninit>, exit=0] -> [exit=0, exit=0]` is actually correct, but I'm nervous about synthesizing the info there.

The code here is supposed to handle `[..., <uninit>] -> [..., exit=137]`

Reply:
ah, that makes sense. I think the larger point is the same as you mentioned in https://github.com/kubernetes/kubernetes/pull/108366#issuecomment-1063278740.

If the init container is not initialized, we don't want to set the exit code to 137. We should only set exit code to 137 if we have confirmed that the container has actually started. As a result with this change, now, e.g. if you have three init containers and first succeeds, second fails, the third will stay on "waiting" as opposed to having exit code 137 set.